delegate-app/HydraAuction/Delegate/Server.hs

```diff
@@ -0,0 +1,131 @@
+
+-- | The config for the delegate server
+data DelegateServerConfig = DelegateServerConfig
+  { host :: IP
+  -- ^ the host of the delegate server
+  , port :: PortNumber
+  -- ^ the port number the delegate server receives input at
+  , tick :: Int
+  -- ^ the amount of milliseconds, the thread should wait
+  , ping :: Int
+  -- ^ the amount of seconds between pings for the client thread
+  }
+
+-- | Representation of the Delegate Server's log
+data DelegateServerLog
+  = Started PortNumber
+  | FrontendConnected ClientId
+  | FrontendInput FrontendRequest
+  | DelegateOutput DelegateResponse
+  | DelegateError DelegateError
+  | ThreadEvent ThreadEvent ThreadSort
+  | QueueAuctionPhaseEvent QueueAuctionPhaseEvent
+  deriving stock (Eq, Show, Generic)
+
+instance Pretty DelegateServerLog where
+  pretty = \case
+    Started port -> "Started Server at Port" <+> viaShow port
+    FrontendConnected clientId ->
+      "Frontend with clientId " <> viaShow clientId <> "connected to Server"
+    DelegateOutput out -> "Delegate output" <> extraInfo (viaShow out)
+    FrontendInput inp -> "Frontend input" <> extraInfo (viaShow inp)
+    DelegateError err -> "Delegate runner error occured" <> extraInfo (pretty err)
+    ThreadEvent ev info -> "Thread" <+> pretty info <> ":" <> extraInfo (pretty ev)
+    QueueAuctionPhaseEvent ev -> "Auction phase queueing" <> extraInfo (pretty ev)
+
+-- | The event type to occur
+data ThreadEvent
+  = ThreadStarted
+  | ThreadCancelled
+  deriving stock (Eq, Ord, Show)
+
+instance Pretty ThreadEvent where
+  pretty = \case
+    ThreadStarted -> "Thread started"
+    ThreadCancelled -> "Thread cancelled"
+
+-- | Which specific thrad the event originates from
+data ThreadSort
+  = WebsocketThread
+  | DelegateRunnerThread
+  | QueueAuctionStageThread
+  deriving stock (Eq, Ord, Show)
+  deriving (Pretty) via ViaShow ThreadSort
+
+-- | an event happening in the queueAuctionPhase thread
+newtype QueueAuctionPhaseEvent
+  = ReceivedAuctionSet AuctionTerms
+  deriving stock (Eq, Ord, Show)
+
+instance Pretty QueueAuctionPhaseEvent where
+  pretty = \case
+    ReceivedAuctionSet terms -> "Received an AuctionSet" <> extraInfo (viaShow terms)
+
+{- | the error that can be thrown by the delegate server, before entering the Delegate
+   transformer
+-}
+newtype DelegateError = FrontendNoParse String
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (FromJSON, ToJSON)
+
+instance Pretty DelegateError where
+  pretty = \case
+    FrontendNoParse err -> "Could not parse the input provided by the frontend" <> extraInfo (pretty err)
+
+-- | trace a 'DelegateServerLog'
+type DelegateTracerT = TracerT DelegateServerLog
+
+-- | like @ServerApp@ but can be used with a transformer
+type ServerAppT m = PendingConnection -> m ()
+
+-- | a newtype for deriving Pretty via Show
+newtype ViaShow a = ViaShow {unViaShow :: a}
+  deriving stock (Eq, Ord, Generic)
+
+instance Show a => Pretty (ViaShow a) where
```

Comment:
I guess the issue is if we say `Show a => Pretty a` then we make instance resolution non-deterministic... but why can't GHC pick a `Pretty` instance if it has one in scope, and only if it can't find one then follow the `Show` branch?

Reply:
Because you don’t want that. This is the motivation for deriving via. You don’t want to have catch all instances because of the open world assumption of Haskell classes.

Reply:
> but why can't GHC pick a Pretty instance if it has one in scope, and only if it can't find one then follow the Show branch?

Is not that what the OVERLAPPABLE does?

> newtype ViaShow

It is strange to me, that Pretty does not already have that implemented. Probably cuz `derviving via` is a recent thing.

Reply:
The good thing is that the whole point of deriving via is that we do not have to care for whether it's implemented already, we can adhoc deriveable instances by just defining a newtype wrapper